phpBB/phpbb/storage/controller/attachment.php

```diff
@@ -0,0 +1,333 @@
+
+				$sql = 'SELECT forum_id, post_visibility
+					FROM ' . POSTS_TABLE . '
+					WHERE post_id = ' . (int) $attachment['post_msg_id'];
+				$result = $db->sql_query($sql);
+				$post_row = $db->sql_fetchrow($result);
+				$db->sql_freeresult($result);
+
+				if (!$post_row || ($post_row['post_visibility'] != ITEM_APPROVED && !$auth->acl_get('m_approve', $post_row['forum_id'])))
+				{
+					// Attachment of a soft deleted post and the user is not allowed to see the post
+					throw new http_exception(404, 'ERROR_NO_ATTACHMENT');
+				}
+			}
+			else
+			{
+				// Attachment is in a private message.
+				$post_row = array('forum_id' => false);
+				phpbb_download_handle_pm_auth($db, $auth, $this->user->data['user_id'], $attachment['post_msg_id']);
+			}
+
+			$extensions = array();
+			if (!extension_allowed($post_row['forum_id'], $attachment['extension'], $extensions))
+			{
+				throw new http_exception(403, sprintf($this->user->lang['EXTENSION_DISABLED_AFTER_POSTING'], $attachment['extension']));
+			}
+		}
+
+		$attachment['physical_filename'] = utf8_basename($attachment['physical_filename']); //?
+
+
+		$download_mode = (int) $extensions[$attachment['extension']]['download_mode'];
+		$display_cat = $extensions[$attachment['extension']]['display_cat'];
+
+		if (($display_cat == ATTACHMENT_CATEGORY_IMAGE || $display_cat == ATTACHMENT_CATEGORY_THUMB) && !$user->optionget('viewimg'))
+		{
+			$display_cat = ATTACHMENT_CATEGORY_NONE;
+		}
+
+		if ($display_cat == ATTACHMENT_CATEGORY_FLASH && !$user->optionget('viewflash'))
+		{
+			$display_cat = ATTACHMENT_CATEGORY_NONE;
+		}
+
+		if ($thumbnail)
+		{
+			$attachment['physical_filename'] = 'thumb_' . $attachment['physical_filename'];
+		}
+		else if ($display_cat == ATTACHMENT_CATEGORY_NONE && !$attachment['is_orphan'] && !phpbb_http_byte_range($attachment['filesize']))
+		{
+			// Update download count
+			phpbb_increment_downloads($db, $attachment['attach_id']);
+		}
+
+		if ($display_cat == ATTACHMENT_CATEGORY_IMAGE && $mode === 'view' && (strpos($attachment['mimetype'], 'image') === 0) && (strpos(strtolower($user->browser), 'msie') !== false) && !phpbb_is_greater_ie_version($user->browser, 7))
+		{
+			wrap_img_in_html(append_sid($phpbb_root_path . 'download/file.' . $phpEx, 'id=' . $attachment['attach_id']), $attachment['real_filename']);
+			file_gc();
+		}
+		else
+		{
+			// Determine the 'presenting'-method
+			if ($download_mode == PHYSICAL_LINK)
+			{
+				// This presenting method should no longer be used
+				if (!@is_dir($phpbb_root_path . $config['upload_path']))
+				{
+					send_status_line(500, 'Internal Server Error');
+					trigger_error($user->lang['PHYSICAL_DOWNLOAD_NOT_POSSIBLE']);
+				}
+
+				redirect($phpbb_root_path . $config['upload_path'] . '/' . $attachment['physical_filename']);
+				file_gc();
+			}
+			else
+			{
+				send_file_to_browser($attachment, $config['upload_path'], $display_cat);
+				file_gc();
```

Comment:
look at the class BinaryResponse. It takes the path of the file and then handles everything